### DISCUSSION TITLE
Installation Troubles

### DISCUSSION THREAD

**Discussion by Procambarus-gracilis** [deepjavalibrary/djl#439] [General]
I am trying to install djl on a linux machine (all command line, no GUI) and I have both Java 9 and git ready to go.

Would someone please explain to me how I am supposed to install djl with this setup?

Unfortunately, my experience with using third party libraries has only been downloading a jar file and using it directly in my code that way, so I am unfamiliar with using libraries by building them from github.  If it makes any difference, I also installed Gradle and it should be ready to go (I've never used it before but I have been reading about it and it seems to be properly installed).  Any help with this would be greatly appreciated.

**Comment by frankfliu** [comment]
@Procambarus-gracilis 
DJL is a set of regular java library. You can download jar files and use in your project without any issue. What you need to do is find out all necessary jars.

The run a DL model, you at least need [api](https://search.maven.org/search?q=a:api), one of engine (MXNet, TF, PT, ONNX)  and it's native library. You also need DJL's dependencies as well. See:
http://docs.djl.ai/master/docs/development/dependency_management.html

It's not straight forward to list all jar files required, that's why most java project use maven or gradle to manage project dependencies.

Another challenge is version matching issue. DJL contains multiple packages, you have to use compatible version together to make it work. We recommend use BOM: http://docs.djl.ai/master/bom/index.html

Can can take a look our example project's gradle (https://github.com/awslabs/djl/blob/master/examples/build.gradle#L11-L36) or maven (https://github.com/awslabs/djl/blob/master/examples/pom.xml#L24-L78) dependencies.

**Comment by Procambarus-gracilis** [comment]
Hi Frank,

Thanks for these pointers.  Since I am a complete novice with all this stuff (and looking at the links you sent me, it is a lot of material), where do you recommend I start?  Should I start learning about gradle first (it looks like a course all by itself!)?  Just looking at the BOM reference, it would appear that everything you mentioned (BOM, dependencies, selecting a native engine) all appear to be specified in the gradle.build file.

Can all of the dependencies, native engine, etc... you mentioned be entirely taken care of through gradle?

My apologies if the question may be elementary, but this is my first time working with such a build system.  Thanks for the link to your gradle file, this looks like it will be very helpful.

Sent from Outlook<http://aka.ms/weboutlook>

________________________________

**Reply by lanking520** [reply]
I would recommend to take a look at here: https://github.com/aws-samples/djl-demo/blob/master/spark/spark3.0/image-classification/build.gradle#L14-L21. (This project can run out-of-box) A bom is a collection model versioning in one place.

**Comment by frankfliu** [comment]
If you just want to create a java project, maven is a bit simpler. Gradle is more powerful, but has higher learning curve.
